File: php/src/LinkedList/LinkedList.php
<?php

class Node
{
    public $e;
    public $next;

    public function __construct($e = null, $next = null)
    {
        $this->e = $e;
        $this->next = $next;
    }

    public function __toString()
    {
        return (string)$this->e;
    }
}


class LinkedList
{
    /** @var Node */
    private $dummyHead;
    private $size;

    public function __construct()
    {
        $this->dummyHead = new Node();
        $this->size = 0;
    }

    public function __toString()
    {
        $str = '';
        $head = $this->dummyHead;
        while ($head->next != null) {
            $head = $head->next;
            $str .= $head;
        }
        return $str;
    }

    public function getSize()
    {
        return $this->size;
    }

    public function isEmpty()
    {
        return $this->size == 0;
    }

    public function add($index, $e)
    {
        if ($index < 0 || $index > $this->size) {
            throw new \Exception('illegal index');
        }

        $prev = $this->dummyHead;
        for ($i = 0; $i < $index; $i++) {
            $prev = $prev->next;
            $prev->next = new Node($e, $prev->next);
            $this->size++;
        }
    }

    public function addFirst($e)
    {
        $this->add(0, $e);
    }

    public function addLast($e)
    {
        $this->add($this->size, $e);
    }
}
//
//$linkedList = new LinkedList();
//$linkedList->addFirst(23);
//$linkedList->addLast(25);
//$linkedList->add(1, 24);
//
//echo $linkedList;

File: LinkedList/LinkedList.go
package LinkedList

type Node struct {
	e interface{}
	next *Node
}

func NewNode(e interface{}, next *Node) *Node {
	return &Node{e: e, next: next}
}

func NewNodeDefault(e interface{}) *Node {
	return NewNode(e,nil)
}

type LinkedList struct {
	head *Node
	size int
}

func (this *LinkedList) GetSize() int {
	return this.size
}

func (this *LinkedList) IsEmpty() bool {
	return this.size == 0
}

func (this *LinkedList) AddFirst(e interface{}) {
	node := NewNodeDefault(e)
	node.next = this.head
	this.head = node
	this.size ++
}

func (this *LinkedList) Add(e interface{}, index int) {
	if index < 0 || index > this.size {
		panic("Add failed, Illegal index.")
	}
	if index == 0 {
		this.AddFirst(e)
	}else {
		var prev = this.head
		for i := 0; i < index - 1; i++ {
			prev = prev.next
		}
		node := NewNodeDefault(e)
		node.next = prev.next
		prev.next = node
		this.size++
	}
}

func (this *LinkedList) AddLast(e interface{}) {
	this.Add(e,this.size)
}

func NewLinkedList() *LinkedList {
	return &LinkedList{nil,0}
}




